hbase_sql/src/main/java/com/splicemachine/derby/stream/spark/SparkTableChecker.java

```diff
@@ -148,30 +148,58 @@ public void setTableDataSet(DataSet tableDataSet) {
-                .combineByKey(new createCombiner(), new mergeValue(), new mergeCombiners())
-                .filter(new DuplicateIndexFilter());
+    private List<String> checkDuplicateIndexes(PairDataSet table, PairDataSet index) throws StandardException {
+        try {
+            SpliceSpark.pushScope(String.format("Check duplicates in index %s.%s", schemaName, indexName));
+            JavaPairRDD duplicateIndexRdd = ((SparkPairDataSet) index).rdd
+                    .combineByKey(new CreateCombiner(), new MergeValue(), new MergeCombiners())
+                    .filter(new DuplicateIndexFilter());
 
-        long count = 0;
-        JavaFutureAction<List> futureAction = duplicateIndexRdd.takeAsync(maxCheckTableError);
-        List<Tuple2<String, List<byte[]>>> result = futureAction.get();
-        for(Tuple2<String, List<byte[]>> tuple : result) {
-            List<byte[]> keys = tuple._2;
-            for (byte[] key : keys) {
-                indexKeyDecoder.set(key, 0, key.length);
-                indexKeyDecoder.decode(indexKeyTemplate);
-                messages.add(indexKeyTemplate.getClone().toString() + "=>" + tuple._1);
-                count++;
+            JavaPairRDD joinedRdd = duplicateIndexRdd
+                    .join(((SparkPairDataSet) table).rdd);
+
+            JavaRDD duplicateIndex = joinedRdd
+                    .mapPartitions(new SparkFlatMapFunction<>(new DeleteDuplicateIndexFunction<>(conglomerate, txn, tentativeIndex, baseColumnMap, fix)));
+
+            Iterator it = duplicateIndex.toLocalIterator();
+
+            if (fix) {
+                return fixDuplicateIndexes(it);
+            } else {
+                return reportDuplicateIndexes(it);
             }
+        }catch (Exception e) {
+            throw StandardException.plainWrapException(e);
         }
-        if (count >= maxCheckTableError) {
-            count = duplicateIndexRdd.count();
-            messages.add("...");
+    }
+
+
+    private List<String> fixDuplicateIndexes(Iterator it) {
+        List<String> messages = Lists.newLinkedList();
+
+        long count = 0;
+        while (it.hasNext()) {
+            Tuple2<String, Tuple2<byte[], ExecRow>> t = (Tuple2<String, Tuple2<byte[], ExecRow>>)it.next();
+            messages.add(t._2._2 + "=>" + t._1);
+            count++;
         }
+        messages.add(0,String.format("Deleted the following %d duplicate indexes", count));
+        return messages;
+    }
+
+    private List<String> reportDuplicateIndexes(Iterator it) throws StandardException, InterruptedException, ExecutionException {
+        List<String> messages = Lists.newLinkedList();
 
+        long count = 0;
+        while (it.hasNext()) {
+            Tuple2<String, Tuple2<byte[], ExecRow>> t = (Tuple2<String, Tuple2<byte[], ExecRow>>)it.next();
+            messages.add(t._2._2 + "=>" + t._1);
+            if (count >= maxCheckTableError) {
+                messages.add("...");
+                break;
```

Comment:
the only difference i see between these two functions is 
if(count >= maxCheckTableError) . I don't see how fixDuplicateIndexes fixes anything.
In general, we could do 

`private List<String> reportDuplicateIndexes(Iterator it, int _maxCheckTableError)`
then have 
`if (_maxCheckTableError >= 0 || count >= _maxCheckTableError)`

and then do
`return reportDuplicateIndexes(it, fix ? -1 : maxCheckTableError);`

Reply:
i think fixDuplicateIndexes is missing the lines https://github.com/splicemachine/spliceengine/pull/3463/files#diff-49eff1a8a3acfe993feb4e304ff615ceL161-L167 , no?

Reply:
Depending on whether fix is true
JavaRDD duplicateIndex = joinedRdd
                    .mapPartitions(new SparkFlatMapFunction<>(new DeleteDuplicateIndexFunction<>(conglomerate, txn, tentativeIndex, baseColumnMap, **fix**)));

DeleteDuplicateIndexFunction deletes (or ignores)  extra indexes